Update player score

beccarothwell opened the pull request:
I've updated the prop names for PlayerScore and provided an updatePlayerName function and input validation for the PlayerScore TextInput

## What's in the PR

### validation folder

validate_player_name.ts 
- exports validatePlayerName function. The function trims trailing whitespace from the provided value, then checks if the trimmed string is empty or contains characters other than alphanumeric or space. Returns an array of error message strings if either validation check fails.

validate_player_name.tests.ts
- provides tests for validatePlayerName

### types folder

player.types.tsx
- changes file ending to .ts as we only need .tsx for files containing jsx i.e. components
- adds PlayerNameAndId type for use in game.ts below

### service folder

game.ts 
- adds getInitialPlayerNames function which takes an array of Player objects and returns an array of simplified PlayerNameAndId objects

### components folder

RoundTrackerPip.tsx
- updates the RoundTrackerPip to include "Round Won", "Round Lost", or "Un-played Round" to tell visually impaired users the status of the round. Uses class "visually-hidden" which will hide the text from users who are not using screen readers

PlayerScore.tsx
- updates the prop names to simplify them, e.g. name instead of playerName

I considered updating the PlayerScore to take a full Player object, but it doesn't need things like the player's cards, or whether the player is human etc. I think it's better to only pass the PlayerScore component the props it needs from each Player, i.e. the name, id, and score.

### src folder

App.tsx 
- adds a new state called inputPlayerNames, and calls getInitialPlayerNames with the initialised game.players to get initial state values
- adds updatePlayerName function which takes an id and a value. The function calls setInputPlayerNames to store the new player name input. It then calls validatePlayerName on the value to validate the input player name. If the value is valid, it calls setGame to add the player name value to the Player with a matching id, trimming trailing whitespace from the value first.
- updates the PlayerScore component name prop to find the player name in inputPlayerNames by matching id
- temporarily displays the player name from the game object to demonstrate that invalid names are not saved to the Game object or displayed elsewhere

Doing it this way means that the TextInput in PlayerScore takes and updates the value in inputPlayerNames. The player can type anything they want in the input field, and it will show error messages if the input is invalid. However the name in the game object will only be updated if the input is valid. This means that when we display the name elsewhere, for example in the Winner component, it will display the last valid name input, and will not save invalid characters. This could be useful in future if we wanted to save the player name to a database.

beccarothwell commented:
Additional note while I'm doing styling, do we want to maybe set a max validation limit on player name size?

jennysivapalan commented:
> Additional note while I'm doing styling, do we want to maybe set a max validation limit on player name size?

Yeah I think so - maybe 15 or 20?

beccarothwell commented:
> > Additional note while I'm doing styling, do we want to maybe set a max validation limit on player name size?
> 
> Yeah I think so - maybe 15 or 20?

Okay I've added validation to limit the name to 15 characters, and also updated the trailing whitespace error warnings

jennysivapalan commented on client/src/App.tsx:
```diff
@@ -162,19 +163,29 @@ function App() {
         roundWinners={roundWinners}
       />
+      <script></script>
+      Player 1 name in game object is: {player1.name}
```
This line can be removed?

beccarothwell replied:
Ah yes, the Player 1 name etc was just for demonstrating that the name was stored differently in the game state. The script I think was an accidental copy paste. Will remove and push again now.

beccarothwell replied:
Done

jennysivapalan commented on client/src/App.tsx:
```diff
@@ -1,60 +1,61 @@
+        };
+      });
+    }
+  };
```
Out of interest could all this login happen in PlayerScore - like the update and validation? Then there's less in App.tsx?

jennysivapalan replied:
Nice let me know if you want me to review after.

beccarothwell replied:
@jennysivapalan I've moved all the validation logic into PlayerScore, it also now has the separate state for the input name before setting it in the game state, good idea on that one. The rest of updatePlayerName I'm pretty sure needs to stay in App.tsx as it's calling setGame to update the game state?

jennysivapalan replied:
Makes sense

jennysivapalan commented on client/src/App.tsx:
```diff
@@ -186,6 +197,7 @@ function App() {
       <p>Player 1 score is: {player1.score}</p>
       <p>Player 1 card is: {player1.cards[0].name}</p>
       <p>Player 1 card is shown: {player1.isCardShown ? "true" : "false"}</p>
+      Player 2 name in game object is: {player2.name}
```
This one can be removed too :)

beccarothwell replied:
Done

jennysivapalan reviewed (approved):
Code wise looks good. I tried to pull the branch but conflicts with App.tsx and I thought no thank you haha.

jennysivapalan commented:
> Code wise looks good. I tried to pull the branch but conflicts with App.tsx and I thought no thank you haha.

not sure why I got conflicts though looks like it should merge ok with main
